src/main/scala/io/qbeast/spark/index/SparkAutoIndexer.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2021 Qbeast Analytics, S.L.
+ */
+package io.qbeast.spark.index
+
+import io.qbeast.core.model.AutoIndexer
+import org.apache.spark.ml.feature.OneHotEncoder
+import org.apache.spark.ml.feature.StringIndexer
+import org.apache.spark.ml.feature.VectorAssembler
+import org.apache.spark.ml.linalg.Matrix
+import org.apache.spark.ml.stat.Correlation
+import org.apache.spark.ml.Pipeline
+import org.apache.spark.qbeast.config.MAX_NUM_COLUMNS_TO_INDEX
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.functions.unix_timestamp
+import org.apache.spark.sql.DataFrame
+
+object SparkAutoIndexer extends AutoIndexer[DataFrame] with Serializable {
+
+  /**
+   * The maximum number of columns to index.
+   *
+   * @return
+   */
+  override def MAX_COLUMNS_TO_INDEX: Int = MAX_NUM_COLUMNS_TO_INDEX
+
+  /**
+   * Chooses the columns to index without limiting the number of columns.
+   *
+   * @param data
+   *   the data to index
+   * @return
+   */
+  override def chooseColumnsToIndex(data: DataFrame): Seq[String] =
+    chooseColumnsToIndex(data, MAX_COLUMNS_TO_INDEX)
+
+  override def chooseColumnsToIndex(data: DataFrame, numColumnsToIndex: Int): Seq[String] = {
```

Comment:
The method is too long to my taste. I would decompose it into a series of calls using the comments you put as the names of the submethods. Again it is a matter of taste.

Reply:
Mm, I would take a look. I would also check if some of the methods can be generalized in the `trait`.

Reply:
I do not have to extract a trait, i.e. introduce a new abstraction, I mean just to extract a few functions to be implemented in the same class next to the mentioned method.

Reply:
I've tried to separate the code in 3 methods:

- withUnixTimestamp (converts columns with timestamp to unix_timestamp)
- withPreprocessedPipeline (which executes the ML Pipeline and returns the DataFrame with correlation matrix)
- getTopNCorrelatedColumns (selects the top N correlated columns.)